echopype/calibrate/calibrate_base.py

```diff
@@ -1,19 +1,216 @@
+
+        Parameters
+        ----------
+        env_params : xr.Dataset
+            The environmental parameters to use for calibration. This data will be interpolated with
+            a provided `EchoData` object.
+
+            When `data_kind` is `"stationary"`, env_params must have a coordinate `"ping_time"`.
+            When `data_kind` is `"mobile"`, env_params must have coordinates `"latitude"`
+            and `"longitude"`.
+            When `data_kind` is `"mobile"`, env_params must have coordinates `"time"`,
+            `"latitude"`, and `"longitude"`.
+        data_kind : {"stationary", "mobile", "organized"}
+            The type of the environmental parameters.
+
+            `"stationary"`: environmental parameters from a fixed location
+            (for example, a single CTD).
+            `"mobile"` environmental parameters from a moving location (for example, a ship).
+            `"organized"`: environmental parameters from many fixed locations
+            (for example, multiple CTDs).
+        interp_method: {"linear", "nearest", "zero", "slinear", "quadratic", "cubic"}
+            Method for interpolation of environmental parameters with the data from the
+            provided `EchoData` object.
+        extrap_method: {"linear", "nearest"}
+            Method for extrapolation of environmental parameters with the data from the
+            provided `EchoData` object. Currently only supported when `data_kind` is `"stationary"`.
+
+        Notes
+        -----
+        Currently cases where `data_kind` is `"organized"` are not supported; support will be added
+        in a future version.
+
+        Examples
+        --------
+        >>> env_params = xr.open_dataset("env_params.nc")
+        >>> EnvParams(env_params, data_kind="mobile", interp_method="linear")
+        >>> echopype.calibrate.compute_Sv(echodata, env_params=env_params)
+        """
+        if interp_method not in VALID_INTERP_METHODS[data_kind]:
+            raise ValueError(
+                f"invalid interp_method {interp_method} for data_kind {data_kind}"
+            )
+
+        self.env_params = env_params
+        self.data_kind = data_kind
+        self.interp_method = interp_method
+        self.extrap_method = extrap_method
+
+    def _apply(self, echodata) -> Dict[str, xr.DataArray]:
+        if self.data_kind == "stationary":
+            dims = ["ping_time"]
+        elif self.data_kind == "mobile":
+            dims = ["latitude", "longitude"]
+        elif self.data_kind == "organized":
+            dims = ["time", "latitude", "longitude"]
+        else:
+            raise ValueError("invalid data_kind")
+
+        env_params = self.env_params
+
+        if self.data_kind == "mobile":
+            # compute_range needs indexing by ping_time
+            interp_plat = echodata.platform.interp(
+                {"location_time": echodata.platform["ping_time"]}
+            )
```

Comment:
This should use `ping_time` from the Beam group since ultimately we want the env variables to be aligned with the acoustic data, not the GPS data.